### ISSUE TITLE
Measuring and improving callback plugin performance

### ISSUE THREAD

**Issue by dmsimard** [ansible-community/ara#171]
## What is the idea ?

There is a performance overhead between running Ansible with and without ara the [ara callback](https://github.com/ansible-community/ara/blob/master/ara/plugins/callback/ara_default.py) because for every hook, instead of doing nothing, Ansible does this:
![doc/source/_static/graphs/recording-workflow.png](https://github.com/ansible-community/ara/raw/master/doc/source/_static/graphs/recording-workflow.png)

Doing something, even if it's very fast, takes longer than doing nothing... because it adds up when the cost is multiplied by thousands of times across the execution of a larger playbook.
So there will always be an overhead (sorry!) but it should be low enough that it's hopefully worth it to the humans :)

The main issue is that the overhead is exponential due to the multiplication of hosts and tasks.
This means at a small scale it will be low enough to be negligible but a real performance hit when running playbooks involving hundreds of hosts or thousands of tasks.

I'm sure there are improvements we could do to reduce this overhead in the current callback, the API clients or the API server and come up with other implementations with fancy stuff like async, threads, or even dropping the events into a local message bus for lowering the latency to a maximum.

Before we start optimizing stuff, I'd like to measure the overhead even if just for the sake of data, metrics and science.

I think a good start would be:

- raw ansible-playbook
- ansible-playbook with 1.x ara callback enabled
- ansible-playbook with 0.x ara callback enabled (for science and comparison purposes, 0.x doesn't have an API server)

There is a similar issue for the measure API server performance here: https://github.com/ansible-community/ara/issues/170

**Comment by srgvg** [comment]
- remote run with ara disabled: total 63s                                                                           
- remote run with ARA enabled: total 156.55s                                                                        
- local run with ARA enabled: total      118.63s                                                                    
remote being from my laptop at office, local being from a vm in the same network as the targets (cloud DC)        
round trip time from laptop to dc is around 15-20ms

**Comment by dmsimard** [comment]
I'll figure out a synthetic benchmark playbook that we can test with more easily but for the time being, I've tested with two different production-ish playbooks.

The main takeaway that I see for now is that even a 15ms latency between the callback and the API server slows things down quite a bit when compared to a locally-running server -- unless the difference can be attributed to sqlite vs mysql, something that would be easier for me to test with a purpose-built playbook.

The remote server:
- version 1.5.1 (latest stable)
- nginx frontend
- gunicorn backend (``gunicorn --workers=9 --bind 127.0.0.1:8000 ara.server.wsgi``)
- mariadb 10.3.17 server on the same node without much particular tweaking

The local server:
- version 1.5.1 (latest stable)
- built-in development server (``ara-manage runserver``)
- sqlite

| playbook | hosts | plays | tasks | results | duration (remote) | duration (local) |
|----------|-------|-------|-------|---------|-------------------|------------------|
| first    | 22    | 74    | 625   | 2754    | 14m50s            | 8m31s            |
| second   | 21    | 114   | 634   | 6035    | 18m57s            | 8m29s            |

I tried running with and without an [experimental patch that adds threads](https://review.opendev.org/#/c/749378/3) but got pretty much the same results.

**Comment by dmsimard** [comment]
I didn't find the time to look at the threads patch but I proposed a simple benchmarking playbook: https://review.opendev.org/#/c/758267/1/tests/integration/benchmark.yaml

For example, running with ``ARA_API_CLIENT=http`` and ``ARA_API_SERVER=https://api.demo.recordsansible.org`` (60ms latency):
```bash
> ara playbook list --path benchmark
+------+-----------+-----------------------------------------+-------+---------+-------+-----------------------------+-----------------+
|   id | status    | path                                    | tasks | results | hosts | started                     | duration        |
+------+-----------+-----------------------------------------+-------+---------+-------+-----------------------------+-----------------+
| 1157 | completed | ...ara/tests/integration/benchmark.yaml |    52 |    1276 |    26 | 2020-10-14T21:01:38.675407Z | 00:02:10.028298 |
+------+-----------+-----------------------------------------+-------+---------+-------+-----------------------------+-----------------+
```

But when running with ``ARA_API_CLIENT=http`` against a server on localhost:
```bash
> ara playbook list --path benchmark
+-----+-----------+-------------------------------------------+-------+---------+-------+----------------------------------+-----------------+
|  id | status    | path                                      | tasks | results | hosts | started                          | duration        |
+-----+-----------+-------------------------------------------+-------+---------+-------+----------------------------------+-----------------+
| 214 | completed | ...ara/tests/integration/benchmark.yaml   |    52 |    1276 |    26 | 2020-10-14T20:45:34.098948-04:00 | 00:00:54.839319 |
+-----+-----------+-------------------------------------------+-------+---------+-------+----------------------------------+-----------------+
```

I think we can use this to gather standardized performance metrics by tweaking the number of hosts and tasks -- we can then measure if whether or not we are improving things with changes (or making them worse!).

Since the data is standard, we should also be able to observe if we have wildly different results for things like the version of ansible or python.

**Comment by andrask** [comment]
Would it be possible to make the reporting asynchronous?

**Comment by dmsimard** [comment]
> Would it be possible to make the reporting asynchronous? That is, send data to the API server asynchronously and let the playbook run on its own pace.

I would like it to be possible, but I am not sure how yet.

One of the challenges is that Ansible callback interface is synchronous in the sense that for each event, Ansible will send a hook to the callback plugin and will continue only when the callback returns like explained in the graph at the beginning of the issue.
So, the longer the callback takes to execute the hook, the longer Ansible will wait and the longer the playbook will take.

The objective is essentially to make each callback hook return as quickly as possible.

 There are probably different approaches to explore, like:
1) threading (ex: a naive [threading patch](https://review.opendev.org/#/c/749378/3/ara/plugins/callback/ara_default.py) but it doesn't seem to work like I thought it would)
2) proper async in-callback (asyncio ? requests-async ?)
3) async through other means (like dropping events in a queue or message bus and processing them asynchronously from there)
4) buffering locally and sending it batched at the end of the playbook

Knowing that the overhead is significantly smaller when running against a server localhost, option 4 could be worth looking into -- despite the [distributed sqlite backend](https://ara.readthedocs.io/en/latest/distributed-sqlite-backend.html) being /almost/ that but it doesn't aggregate data into a single database, it keeps the data separate.

Very much open to ideas, suggestions and contributions :)

**Comment by andrask** [comment]
Disclaimer: I haven't checked the implementation deeply yet.

In my case I ran several playbooks in series. I used Mitogen and Ara at the same time and everything I gained with Mitogen (~31% speed increase) was diminished by Ara. I confirmed this by turning off Ara.

Yes, so the important thing is to make the callback a fire and forget event emitter. 4 is a real alternative, for sure, though we would lose the cool feature of real-time state.

With the other options, some other entity, a proxy, should be responsible for transferring the events to the API server. If we can rely on Linux and POSIX entities, then MessageQueues are one option, or UDP sockets are also viable for the IPC between the processes (and there are tons of other solutions, too). The proxy component may be started smartly, like the ControlPath in ssh. 

I also looked at your threading patch. It seems like it doesn't do what you really intended. I mean like here https://review.opendev.org/#/c/749378/3/ara/plugins/callback/ara_default.py@351

    def v2_runner_on_ok(self, result, **kwargs):
        self.threads.submit(self._load_result(result, "ok", **kwargs))

The submit will submit the result of the _load_result, which is always None. But the _load_result is executed synchronously, so nothing changed. Changing the call to 

        self.threads.submit(self._load_result, result, "ok", **kwargs)

may be better. It will cause the function to execute asynchronously. I also commented on the review.

The other thing I found though is that the GUI seems to be a little broken as search for host and task never work for me. Always returned everything. Honestly, I feel like pouring all this data directly to Elasticsearch and visualizing it with Kibana may be a better solution. It allows for much more analysis out of the box.

**Comment by dmsimard** [comment]
> In my case I ran several playbooks in series. I used Mitogen and Ara at the same time and everything I gained with Mitogen (~31% speed increase) was diminished by Ara. I confirmed this by turning off Ara.

Indeed, not surprising.

> Yes, so the important thing is to make the callback a fire and forget event emitter. 4 is a real alternative, for sure, though we would lose the cool feature of real-time state.

Right.

> With the other options, some other entity, a proxy, should be responsible for transferring the events to the API server. If we can rely on Linux and POSIX entities, then MessageQueues are one option, or UDP sockets are also viable for the IPC between the processes (and there are tons of other solutions, too). The proxy component may be started smartly, like the ControlPath in ssh.

Yup. Although given the results of the threading patch (more info below) I'd like to see how far we can take this before resorting to other means.

> I also looked at your threading patch. It seems like it doesn't do what you really intended. I mean like here https://review.opendev.org/#/c/749378/3/ara/plugins/callback/ara_default.py@351
> 
> ```
> def v2_runner_on_ok(self, result, **kwargs):
>     self.threads.submit(self._load_result(result, "ok", **kwargs))
> ```
> 
> The submit will submit the result of the _load_result, which is always None. But the _load_result is executed synchronously, so nothing changed. Changing the call to
> 
> ```
>     self.threads.submit(self._load_result, result, "ok", **kwargs)
> ```
> 
> may be better. It will cause the function to execute asynchronously. I also commented on the review.

That's amazing, it works and almost brings a tear to my eye... thank you so much :tada: 
```
> ara playbook list --path benchmark
+------+-----------+-----------------------------------------+-------+---------+-------+-----------------------------+-----------------+
|   id | status    | path                                    | tasks | results | hosts | started                     | duration        |
+------+-----------+-----------------------------------------+-------+---------+-------+-----------------------------+-----------------+
| 1179 | completed | ...ara/tests/integration/benchmark.yaml |    52 |    1265 |    26 | 2020-10-16T09:17:14.864783Z | 00:00:22.148230 |
| 1178 | completed | ...ara/tests/integration/benchmark.yaml |    52 |    1276 |    26 | 2020-10-16T09:07:00.723170Z | 00:00:26.230551 |
| 1177 | completed | ...ara/tests/integration/benchmark.yaml |    52 |    1276 |    26 | 2020-10-16T09:06:24.774177Z | 00:00:26.006876 |
| 1176 | completed | ...ara/tests/integration/benchmark.yaml |    52 |    1276 |    26 | 2020-10-16T08:59:13.020078Z | 00:00:26.553335 |
| 1157 | completed | ...ara/tests/integration/benchmark.yaml |    52 |    1276 |    26 | 2020-10-14T21:01:38.675407Z | 00:02:10.028298 |
+------+-----------+-----------------------------------------+-------+---------+-------+-----------------------------+-----------------+
```

> The other thing I found though is that the GUI seems to be a little broken as search for host and task never work for me. Always returned everything.

Could you please spin this off into a different issue and provide a bit more info on how to reproduce ? If I click on a host in the host table it properly filters things for the specified host, for example: https://api.demo.recordsansible.org/playbooks/1007.html?host=2138&changed=True#results

> Honestly, I feel like pouring all this data directly to Elasticsearch and visualizing it with Kibana may be a better solution. It allows for much more analysis out of the box.

Sure, but then you also need elasticsearch and kibana which is not mutually exclusive, anyway :)
There is a logstash callback here you might find interesting: https://github.com/ansible-collections/community.general/blob/main/plugins/callback/logstash.py

**Comment by andrask** [comment]
Great! That's awesome when things start to work.

The GUI thing is: so it really needs the host id and task id. I was trying to search for task name. :)

**Comment by dmsimard** [comment]
> Great! That's awesome when things start to work.
> 
> The GUI thing is: so it really needs the host id and task id. I was trying to search for task name. :)

Yes, it's somewhat of a limitation of the API -- results are tied to a host ID, not a host name.
We could make search by name work but that's how it is right now.

**Comment by frukto** [comment]
>     4\. buffering locally and sending it batched at the end of the playbook
> 
> 
> Knowing that the overhead is significantly smaller when running against a server localhost, option 4 could be worth looking into -- despite the [distributed sqlite backend](https://ara.readthedocs.io/en/latest/distributed-sqlite-backend.html) being /almost/ that but it doesn't aggregate data into a single database,

The tricky part is to deal with all the UNIQUE constrains with overlapping IDs in distributed/local DBs. But it seems feasible. For option 2, I have no idea if that would be possible have proper async requests to ARA server and let the hooks return immediately.

**Comment by dmsimard** [comment]
@frukto o/

> The tricky part is to deal with all the UNIQUE constrains with overlapping IDs in distributed/local DBs. But it seems feasible.

Yeah, we can't merge the databases easily because the primary keys and other things will clash -- not to mention if there happens to be deltas between database migrations.

I'm not convinced this would be the right approach but something I could see working would be to have a script or a command that iterates through all the playbooks and their results and then posts them to another API server.

> For option 2, I have no idea if that would be possible have proper async requests to ARA server and let the hooks return immediately.

It turns out that threads might be simple and "good enough" once we get around to optimizing them, the callback and the client a bit further but I'm very open to different implementations -- the API is there for that.

For now I'm gathering some benchmark data and I'll post it soon.

**Comment by dmsimard** [comment]
Some benchmark data.

(Work in progress, will edit as data becomes available)

## ara 1.5.3 with sqlite

With the default offline API client against localhost (no running API server):
```
> ara playbook list --label="client:offline"
+-----+-----------+-----------------------------------------+-------+---------+-------+----------------------------------+-----------------+
|  id | status    | path                                    | tasks | results | hosts | started                          | duration        |
+-----+-----------+-----------------------------------------+-------+---------+-------+----------------------------------+-----------------+
| 310 | completed | ...ara/tests/integration/benchmark.yaml |   202 |   40201 |   201 | 2020-10-31T11:30:12.313649-04:00 | 00:29:25.955674 |
| 309 | completed | ...ara/tests/integration/benchmark.yaml |    52 |    5101 |   101 | 2020-10-31T11:24:58.979017-04:00 | 00:03:22.571607 |
| 308 | completed | ...ara/tests/integration/benchmark.yaml |    52 |    1276 |    26 | 2020-10-31T11:22:05.943304-04:00 | 00:00:52.728591 |
+-----+-----------+-----------------------------------------+-------+---------+-------+----------------------------------+-----------------+
```

With ``export ARA_API_CLIENT=http`` against localhost: ``ara-manage runserver``
```
> ara playbook list --label="client:http"
+-----+-----------+-----------------------------------------+-------+---------+-------+----------------------------------+-----------------+
|  id | status    | path                                    | tasks | results | hosts | started                          | duration        |
+-----+-----------+-----------------------------------------+-------+---------+-------+----------------------------------+-----------------+
| 314 | completed | ...ara/tests/integration/benchmark.yaml |   202 |   40201 |   201 | 2020-10-31T12:25:58.707298-04:00 | 00:17:24.335043 |
| 313 | completed | ...ara/tests/integration/benchmark.yaml |    52 |    5101 |   101 | 2020-10-31T12:23:34.120747-04:00 | 00:01:37.972033 |
| 311 | completed | ...ara/tests/integration/benchmark.yaml |    52 |    1276 |    26 | 2020-10-31T12:14:12.684240-04:00 | 00:00:30.101029 |
+-----+-----------+-----------------------------------------+-------+---------+-------+----------------------------------+-----------------+
```

With ``export ARA_API_CLIENT=http`` against localhost: ``gunicorn --access-logfile - --workers=9 --bind 127.0.0.1:8000 ara.server.wsgi``
```
> ara playbook list --label="backend:gunicorn"
+-----+-----------+-----------------------------------------+-------+---------+-------+----------------------------------+-----------------+
|  id | status    | path                                    | tasks | results | hosts | started                          | duration        |
+-----+-----------+-----------------------------------------+-------+---------+-------+----------------------------------+-----------------+
| 317 | completed | ...ara/tests/integration/benchmark.yaml |   202 |   20101 |   101 | 2020-10-31T13:17:05.314580-04:00 | 00:06:18.769420 |
| 316 | completed | ...ara/tests/integration/benchmark.yaml |    52 |    5101 |   101 | 2020-10-31T13:14:17.244041-04:00 | 00:01:09.821960 |
| 315 | completed | ...ara/tests/integration/benchmark.yaml |    52 |    1276 |    26 | 2020-10-31T13:12:09.157927-04:00 | 00:00:20.606147 |
+-----+-----------+-----------------------------------------+-------+---------+-------+----------------------------------+-----------------+
```

**Comment by dmsimard** [comment]
Wrote a blog post about the data: https://ara.recordsansible.org/blog/2020/11/01/benchmarking-ansible-and-ara-for-fun-and-science/

**Comment by andrask** [comment]
Very interesting read and promising results. I may have a few ideas how the improvement could proceed.

### How much is the minimal overhead?
For me, it seems to be a missing piece of the puzzle. You can scrape away as much of the additional operations as possible but what is the minimum overhead of having the callback enabled. That is, something that collects all the data from Ansible and then discards it. It would be really useful to see how much this is compared to the pure Ansible run as this could be base optimal state that we are trying to reach.

### Synchronous calls
One of the preventors of things going as fast as possible is the fact that some API calls are made synchronously (as far as I understood) I guess these are necessary because the server is providing some unique IDs. I'm talking from gut feeling from now on.

So if we could turn the updates toward the server a single direction flow of events, then all communication could essentially be asynchronous. In order to achieve this, the client must provide/generate all IDs for the events. The events can then be streamed to the server and it can process them in order.

Once I had a similar problem where I had a high performance task framework and the state of the tasks had to be represented in a job hierarchy. Originally, the comms was synchronous and it kept the task framework from performing adequately. The only thing I actually needed on the tasks side from the server was the job ID. I created the job and then used the returned ID to send updates to that. But I figured out that this ID can actually be anything, as long as it's unique in the event stream from the client. So I changed the logic so that each task would be assigned a simple 32 bit integer ID that is unique for that client and every event from that task would refer to this ID. Then the events were streamed to the job server where the server took all events in order and it knew that each event from a given client with a given task ID must be applied on the same job object.

I hope this makes sense. 

If I understand right, then [this](https://github.com/ansible-community/ara/blob/3837b3b6e29063072aae818e3d23e40ca9c1499f/ara/plugins/callback/ara_default.py#L337) and [this](https://github.com/ansible-community/ara/blob/3837b3b6e29063072aae818e3d23e40ca9c1499f/ara/plugins/callback/ara_default.py#L304) are the code that need to run synchronously. By applying the logic I described, these could be turned into async calls/events, too.

**Comment by dmsimard** [comment]
> Very interesting read and promising results. I may have a few ideas how the improvement could proceed.

@andrask o/ thanks for your interest

> ### How much is the minimal overhead?
> 
> For me, it seems to be a missing piece of the puzzle. You can scrape away as much of the additional operations as possible but what is the minimum overhead of having the callback enabled. That is, something that collects all the data from Ansible and then discards it. It would be really useful to see how much this is compared to the pure Ansible run as this could be base optimal state that we are trying to reach.

You understand correctly. I've benchmarked "no callback enabled" and "ara callback enabled" but we could argue that there's another delta to measure: what's the overhead cost of a callback that does nothing (or just a print?) so we could have the true overhead of the ara callback.

For what it's worth, Zuul has a [callback plugin](https://opendev.org/zuul/zuul/src/branch/master/zuul/ansible/base/callback/zuul_json.py) which saves data into a local json file and there's no database. It could be interesting to compare :)

> ### Eliminating synchronous calls
> 
> One of the preventors of things going as fast as possible is the fact that some API calls are made synchronously (as far as I understood) I guess these are necessary because the server is providing some unique IDs. I'm talking from gut feeling from now on.
> So if we could turn the updates toward the server a single direction flow of events, then all communication could essentially be asynchronous. In order to achieve this, the client must provide/generate all IDs for the events. The events can then be streamed to the server and it can process them in order.

There is some documentation about [relationship between objects](https://ara.readthedocs.io/en/latest/api-documentation.html#relationship-between-objects) but, for example, a play is expected to be associated to a playbook so we need to have a playbook id.

> Once I had a similar problem where I had a high performance task framework and the state of the tasks had to be represented in a job hierarchy. Originally, the comms was synchronous and it kept the task framework from performing adequately. The only thing I actually needed on the tasks side from the server was the job ID. I created the job and then used the returned ID to send updates to that. But I figured out that this ID can actually be anything, as long as it's unique in the event stream from the client. So I changed the logic so that each task would be assigned a simple 32 bit integer ID that is unique for that client and every event from that task would refer to this ID. Then the events were streamed to the job server where the server took all events in order and it knew that each event from a given client with a given task ID must be applied on the same job object.

There can be something else than the callback which ends up sending the data in a different order but with the current API implementation, there is still some order that must be respected so the expected associations can be made.
We can keep this in mind but re-designing the API is out of scope for now :p

> I hope this makes sense.

Yes :+1: 

> If I understand right, then [this](https://github.com/ansible-community/ara/blob/3837b3b6e29063072aae818e3d23e40ca9c1499f/ara/plugins/callback/ara_default.py#L337) and [this](https://github.com/ansible-community/ara/blob/3837b3b6e29063072aae818e3d23e40ca9c1499f/ara/plugins/callback/ara_default.py#L304) are the code that need to run synchronously. By applying the logic I described, these could be turned into async calls/events, too.

The bits that are currently pushed to threads in the callback ( https://codesearch.recordsansible.org/?q=threads.submit ) are because at that point, their "parent" relationship dependency has been met and the order in which they are processed isn't important, only that they are backgrounded and then processed as fast as possible before checkpoints like at the end of each task: https://github.com/ansible-community/ara/blob/3037c16db15f8452f7b061e01332ebff86d4c001/ara/plugins/callback/ara_default.py#L391-L393

Then at the end of the playbook there's a last checkpoint to make sure threads have returned before exiting: https://github.com/ansible-community/ara/blob/3037c16db15f8452f7b061e01332ebff86d4c001/ara/plugins/callback/ara_default.py#L420-L421

**Comment by srgvg** [comment]
>
>
> You understand correctly. I've benchmarked "no callback enabled" and "ara
> callback enabled" but we could argue that there's another delta to measure:
> what's the overhead cost of a callback that does nothing (or just a print?)
> so we could have the true overhead of the ara callback.
>
> For what it's worth, Zuul has a callback plugin
> <https://opendev.org/zuul/zuul/src/branch/master/zuul/ansible/base/callback/zuul_json.py>
> which saves data into a local json file and there's no database. It could
> be interesting to compare :)
>

FWIS, I'd think comparing ARA with a minimal callback would be more useful
than a 'functionally empty' callback.

**Comment by andrask** [comment]
> FWIS, I'd think comparing ARA with a minimal callback would be more useful
than a 'functionally empty' callback.

Having a zero functionality measurement is the base line. There is no way to make Ara better. I think it's fairly important to know this number to be able to compare the current status. This is especially important if we consider that the basic Ansible run is slow compared to, for example, running with Mitogen.

I'm not sure how I can run the benchmark tests. Can you please create a script or tox env that does it?

> their "parent" relationship

Yes, this is exactly what I tried to explain. This relationship is different for the client and the server and both entities can create a consistent view of the world but this doesn't mean that the client has to use the same model as the server. The only thing that has to be consistent is the order of events from the client. I was trying to find something similar for this operational mode and I think the Elasticsearch Bulk APIs are a good example. In the bulk request, individual operations are sent as a package. The Ara client could send its events the same way. I'm not knowledgeable in Django but may try to tweak with the setup a bit.

**Comment by dmsimard** [comment]
> FWIS, I'd think comparing ARA with a minimal callback would be more useful than a 'functionally empty' callback.

I don't see that as mutually exclusive -- test all the things \o/

We could even be surprised by the results :)

**Comment by dmsimard** [comment]
> I'm not sure how I can run the benchmark tests. Can you please create a script or tox env that does it?

I've sent a patch that adds a tox target: https://review.opendev.org/#/c/761409/
You can check it out and run it like this:
```
git clone https://github.com/ansible-community/ara
cd ara
git fetch https://review.opendev.org/recordsansible/ara refs/changes/09/761409/1 && git checkout FETCH_HEAD
tox -e benchmark -- -e benchmark_host_count=50 -e benchmark_task_count=50
```

> Yes, this is exactly what I tried to explain. This relationship is different for the client and the server and both entities can create a consistent view of the world but this doesn't mean that the client has to use the same model as the server. The only thing that has to be consistent is the order of events from the client. I was trying to find something similar for this operational mode and I think the Elasticsearch Bulk APIs are a good example. In the bulk request, individual operations are sent as a package. The Ara client could send its events the same way. I'm not knowledgeable in Django but may try to tweak with the setup a bit.

I'm not a django pro either but there's probably a precedent for this, knowing how many apps of all kind running on top of django.
Let me know if you find anything interesting :)

**Comment by valleedelisle** [comment]
I have a use case where I might be interested in buffering locally, into a file and import later into the API. The ARA API would be running on a totally different network and pushed by some other process after the playbook is completed. How complicated would this be? From my understanding, the callback plugin is just pushing stuff into the database, why doesn't it have to synced?

**Comment by phemmer** [comment]
I wrote a wrapper around the ara callback plugin that can be found here: https://gist.github.com/phemmer/8ee4ea0ebf1b389050ce4a2bd78c66d6  

It basically makes everything asynchronous, preventing the ARA hooks from blocking Ansible execution. It's a single queue and a single thread, so order is preserved.  
Note that because everything is processed in a queue, at the end of the playbook if it hasn't caught up, Ansible will wait for the queue to drain before exiting. This also means that if Ansible is interrupted, ARA might not have all the results.

To use, just remove the override (`ANSIBLE_CALLBACK_PLUGINS="$(python3 -m ara.setup.callback_plugins)"`), and put this in the normal callback directory (`~/.ansible/plugins/callback:/usr/share/ansible/plugins/callback`).
